README.md
# Assay MiniTest

[Homepage](http://rubyworks.github.com/assay-minitest) /
[Source Code](http://github.com/rubyworks/assay-minitest) /
[MailingList](http://groups.google.com/group/rubyworks-mailinglist)


## DESCRIPTION

Assay MiniTest is a compatability layer, built on top of [Assay]([http://rubyworks.github.com/assay),
that makes it possible for developers to use MiniTest compatible assertions
with any BRASS-compliant test framework. This allows developers to
change test frameworks without having to change a slew of previously
defined assertions calls.

Assay MiniTest works by defining MiniTest-compatible assertion methods
which in turn call on Assay's assertion classes. Assay defines assertions
in the same way that Ruby defines exceptions. An assertion is nothing more
that an extended Exception class. Assay provides a complete set of these
assertion classes for all common assertion needs. See [Assay](http://rubyworks.github.com/assay)
project for more information on this foundational library.


## SYNOPSIS

Simply require the `assay/minitest` script, and include the `Assay::MiniTest::Assertions` 
or `Assay::Assertions` mixin module into your tests wherever your test framework
requires it (which  may be as simple as the toplevel namespace).

    require 'assay/minitest'

    include Assay::Assertions

Now assertions can be made just as if you were using MiniTest.

    assert_equal(10, 5+5)

    refute_kind_of(String, 10)

Alternately, you can use MiniTest's expectations notation via `Assay::MiniTest::Extensions` 
or just `Assay::Extensions`.

    include Assay::Extensions

    (5+5).must_equal(10)

    10.must_be_kind_of?(Fixnum)


## LIMITATIONS

Note that compatibility is not 100%, though it is fairly close.
Compatibilty will improve with future releases. Please feel _obligated_
to submit a patch, if you need a missing a feature ;)


## INSTALLATION

To install with RubyGems simply open a console and type:

    $ gem install assay-minitest

Site installation with the tarball can be done with Ruby Setup
(gem install setup). See http://rubyworks.github.com/setup.


## COPYRIGHTS

Copyright (c) 2012 Rubyworks

This program is ditributed under the terms of the *BSD-2-Clause* license.

See LICENSE.rdoc file for details.



HISTORY.md
# HISTORY

## 0.1.0 | 2012-01-26

This is the initial release of Assay MiniTest. Compatibility
with MiniTest's assertion and expectation methods is nearly
complete.

Changes:

* Happy Birthday!



spec/applique/setup.rb
require 'assay/minitest'


lib/assay-minitest/extensions.rb
module Assay; end
module Assay::MiniTest

  # This module provides extensions for MiniTest compatiblity.
  #
  # The set is not fully compataible, but provides most of MiniTest
  # extensions.
  #
  # Compatability will improve with time.
  #
  module Extensions

    # Passes it operator with optional argument returns successfully.
    #
    #   10.must_be :>, 4
    #   10.must_be :even?
    #
    # @raise ExecutionAssay
    #
    def must_be(operator, argument=nil, msg=nil)
      if argument
        ExecutionAssay.assert!(:message=>msg) do
          __send__(operator, argument)
        end
      else
        ExecutionAssay.assert!(:message=>msg) do
          __send__(operator)
        end
      end
    end

    #
    # Passes it operator with optional argument returns negatively.
    #
    #   10.wont_be :>, 20
    #   10.wont_be :odd?
    #
    # @raise ExecutionAssay
    #
    def wont_be(operator, argument=nil, msg=nil)
      if argument
        ExecutionAssay.refute!(:message=>msg) do
          self.__send__(operator, argument)
        end
      else
        ExecutionAssay.refute!(:message=>msg) do
          self.__send__(operator)
        end
      end
    end
  
    #
    # Passes if object is like criterion.
    #
    #   object.must_be_like(criterion)
    #
    # @raise LikeAssay
    #
    def must_be_like(exp, msg=nil)
      LikeAssay.assert!(self, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object is not like criterion using {CompareAssay}.
    #
    #   object.wont_be_like(criterion)
    #
    # @raise LikeAssay
    #
    def wont_be_like(exp, msg=nil)
      LikeAssay.refute!(self, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if expected and actual are equal within absolute delta tolerance.
    #
    #   0.05.must_be_within_delta (50000.0 / 10**6), 0.00001
    #
    # @raise WithinAssay
    #
    def must_be_within_delta(exp, delta, msg=nil)
      WithinAssay.assert!(self, exp, delta, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if expected and actual are NOT equal within absolute delta tolerance.
    #
    #   0.05.wont_be_within_delta (50000.0 / 10**6), 0.00001
    #
    # @raise WithinAssay
    #
    def wont_be_within_delta(exp, delta, msg=nil)
      WithinAssay.refute!(self, exp, delta, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if expected and actual are equal within relative epsilon tolerance.
    #
    #   0.05.must_be_within_epsilon (50000.0 / 10**6), 0.00001
    #
    # @raise CloseAssay
    #
    def must_be_within_epsilon(exp, epsilon, msg=nil)
      CloseAssay.assert!(self, exp, epsilon, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if expected and actual are NOT equal within relative epsilon tolerance.
    #
    #   0.05.wont_be_within_epsilon (50000.0 / 10**6), 0.00001
    #
    # @raise CloseAssay
    #
    def wont_be_within_epsilon(exp, epsilon, msg=nil)
      WithinAssay.refute!(self, exp, epsilon, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object is empty.
    #
    #   object.must_be_empty
    #
    # @raise EmptyAssay
    #
    def must_be_empty(msg=nil)
      EmptyAssay.assert!(self, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object is not empty.
    #
    #   object.wont_be_empty
    #
    # @raise EmptyAssay
    #
    def wont_be_empty(msg=nil)
      EmptyAssay.refute!(self, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if expected == +actual.
    #
    # Note that the ordering of arguments is important,
    # since a helpful error message is generated when this
    # one fails that tells you the values of expected and actual.
    #
    #   'MY STRING'.must_equal('my string'.upcase)
    #
    # @raise EqualAssay
    #
    def must_equal(exp, msg=nil)
      EqualAssay.assert!(self, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if expected != actual
    #
    #  'some string'.wont_equal('some other string')
    #
    # @raise EqualAssay
    #
    def wont_equal(exp, msg=nil)
       EqualAssay.refute!(self, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if method call returns successfully.
    #
    #   "string".must_send([:upcase], "Doesn't upcase!")
    #
    # @raise ExecutionAssay
    #
    def must_send(send_array, msg=nil)
      ExecutionAssay.assert!(:message=>msg, :backtrace=>caller) do
        self.__send__(*send_array)
      end
    end

    #
    # Passes if method call does not return successfully.
    #
    #   "string".wont_send([:upcase], "Damn you, Uppercase!")
    #
    # @raise ExecutionAssay
    #
    def wont_send(send_array, msg=nil)
      ExecutionAssay.refute!(:message=>msg, :backtrace=>caller) do
        self.__send__(*send_array)
      end
    end

    #
    # Passes if the block yields successfully.
    #
    #   5.must_satisfy{ |x| x > 3 }
    #
    # @raise ExecutionAssay
    #
    def must_satisfy(msg=nil, &block)
      ExecutionAssay.assert!(self, :message=>msg, :backtrace=>caller, &block)
    end

    #
    # Passes if the block does not yield successfully.
    #
    #   5.wont_satisfy{ |x| x < 3 }
    #
    # @raise ExecutionAssay
    #
    def wont_satisfy(msg=nil, &block)
      ExecutionAssay.refute!(self, :message=>msg, :backtrace=>caller, &block)
    end

    #
    # Passed if object is +false+.
    #
    # @raise FalseAssay
    #
    def must_be_false(msg=nil)
      FalseAssay.assert!(self, :message=>msg, :backtrace=>caller)
    end

    #
    # Passed if object is not +false+.
    #
    #   assert_not_false(false)
    #
    # @raise FalseAssay
    #
    def wont_be_false(msg=nil)
      FalseAssay.refute!(self, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if `self` is the same exact object as expected.
    #
    #   object.must_be_same_as(object)
    #
    # @raise IdentityAssay
    #
    def must_be_same_as(exp, msg=nil)
      IdentityAssay.assert!(self, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if actual is not the same exact object as expected.
    #
    #   object.wont_be_same_as(other)
    #
    # @raise IdentityAssay
    #
    def wont_be_same_as(exp, msg=nil)
      IdentityAssay.refute!(self, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object is an instance of class.
    #
    #   'foo'.must_be_instance_of(String)
    #
    # @raise InstanceAssay
    #
    def must_be_instance_of(cls, msg=nil)
      InstanceAssay.assert!(self, cls, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object is not an instance of class.
    #
    #   'foo'.wont_be_instance_of(Integer)
    #
    # @raise InstanceAssay
    #
    def wont_be_instance_of(cls, msg=nil)
      InstanceAssay.refute!(self, cls, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object is a kind of class.
    #
    #   assert_kind_of(Object, 'foo')
    #
    # @raise KindAssay
    #
    def must_be_kind_of(cls, msg=nil)
      KindAssay.assert!(self, cls, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object is not a kind of class.
    #
    #   assert_not_kind_of(Object, 'foo')
    #
    # @raise KindAssay
    #
    def wont_be_kind_of(cls, msg=nil)
      KindAssay.refute!(self, cls, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object matches pattern using `#=~` method.
    #
    #   'one 2 three'.must_match(/two/)
    #
    # @raise MatchAssay
    #
    def must_match(exp, msg=nil)
      MatchAssay.assert!(self, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object does not match pattern using `#=~` method.
    #
    #   'one 2 three'.wont_match(/two/)
    #
    # @raise MatchAssay
    #
    def wont_match(exp, msg=nil)
      MatchAssay.refute!(self, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passed if object is +nil+.
    #
    # @raise NilAssay
    #
    def must_be_nil(msg=nil)
      NilAssay.assert!(self, :message=>msg, :backtrace=>caller)
    end

    #
    # Passed if object is not +nil+.
    #
    #   assert_not_nil(true)
    #
    # @raise NilAssay
    #
    def wont_be_nil(msg=nil)
      NilAssay.refute!(self, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if the procedure raises a given exception.
    #
    #   lambda{ raise 'Boom!!!' }.must_raise(RuntimeError)
    #
    # @raise RaiseAssay
    #
    def must_raise(exp, msg=nil, call=nil)
      RaiseAssay.assert!(exp, msg=nil, call=nil, &self)
    end

    #
    # Passes if the procedure *does not* raise a given exceptions.
    #
    #   lambda{ raise 'Boom!!!' }.wont_raise(IOError)
    #
    # @raise RaiseAssay
    #
    def wont_raise(exp, msg=nil, call=nil)
      RaiseAssay.refute!(exp, msg, call, &self)
    end

    #
    # Passes if +object+ respond_to? +methods+.
    #
    #   'bugbear'.must_respond_to(:slice)
    #
    # @raise RespondAssay
    #
    def must_respond_to(method, msg=nil)
      RespondAssay.assert!(self, method, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if +object+ does not respond_to? +methods+.
    #
    #   'bugbear'.wont_respond_to(:slice)
    #
    # @raise RespondAssay
    #
    def wont_respond_to(method, msg=nil)
      RespondAssay.refute!(self, method, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if `criterion.eql?(actual)`.
    #
    # Note that the ordering of arguments is important,
    # since a helpful error message is generated when this
    # one fails that tells you the values of expected and actual.
    #
    #   'MY STRING'.must_be_equivalent_to('my string'.upcase)
    #
    # @raise EqualityAssay
    #
    def must_be_equivalent_to(criterion, msg=nil)
      EqualityAssay.assert!(self, criterion, :message=>msg, :backtrace=>caller)
    end

    alias_method :must_eql, :must_be_equivalent_to

    #
    # Passes if NOT `criterion.eql?(actual)`.
    #
    #   'MY STRING'.wont_be_equivalent_to('some other string')
    #
    # @raise EqualityAssay
    #
    def wont_be_equivalent_to(criterion, msg=nil)
      EqualityAssay.refute!(self, criterion, :message=>msg, :backtrace=>caller)
    end

    alias_method :wont_eql, :wont_be_equivalent_to

    #
    # Passes if the block throws expected_symbol
    #
    #   assert_throws :done do
    #     throw :done
    #   end
    #
    # @raise ThrowAssay
    #
    def must_throw(sym, msg=nil)
      ThrowAssay.assert!(sym, :message=>msg, :backtrace=>caller, &self)
    end

    #
    # Passes if the block throws expected_symbol
    #
    #   refute_throws :done do
    #     throw :chimp
    #   end
    #
    # @raise ThrowAssay
    #
    def wont_throw(sym, msg=nil)
      ThrowAssay.refute!(sym, :message=>msg, :backtrace=>caller, &self)
    end

    #
    # Passed if object is +true+.
    #
    #   object.must_be_true
    #
    # @raise TrueAssay
    #
    def must_be_true(msg=nil)
      TrueAssay.assert!(self, :message=>msg, :backtrace=>caller)
    end

    #
    # Passed if object is not +true+.
    #
    #   object.wont_be_true
    #
    # @raise TrueAssay
    #
    def wont_be_true(msg=nil)
      TrueAssay.refute!(self, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if `self.include?(object)`.
    #
    # @raise IncludeAssay
    #
    def must_include(object, msg=nil)
      IncludeAssay.assert!(self, object, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if `self.include?(object)`.
    #
    # @raise IncludeAssay
    #
    def wont_include(object, msg=nil)
      IncludeAssay.refute!(self, object, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if output matches +pattern+.
    #
    # @example
    #   proc{ puts "fun!" }.must_output('fun!')
    #
    # @raise OutputAssay
    #
    def must_output(pattern, msg=nil)
      OutputAssay.assert!(pattern, :message=>msg, :backtrace=>caller, &self)
    end

  end

end


lib/assay-minitest/assertions.rb
module Assay; end
module Assay::MiniTest

  # This module holds the MiniTest assertion methods for MiniTest
  # compatibility.
  #
  # While it does not provide 100% of MiniTest's assertions at the moment,
  # compatibility is very close and will improved with upcoming releases.
  #
  # TODO: Should we adjust error messages to be like MiniTests ?
  #
  module Assertions

    #
    def assert(truth, msg=nil)
      Assertion.assert!(truth, :message=>msg)
    end

    #
    def refute(untruth, msg=nil)
      Assertion.refute!(untruth, :message=>msg)
    end

    #
    #def assert_alias_method(object, alias_name, original_name, message = nil)
    #end

    #
    # Passes if actual is like expected, where `like` means satisfyin any one 
    # of `#===`, `#==`, `#eql?` or `#equal?` calls.
    #
    # This is not strictly a Test::Unit assertion but is added here to cover
    # all of Assay's availabe assertion classes.
    #
    def assert_alike(exp, act, msg=nil)
      LikeAssay.assert!(act, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if actual is NOT like expected, where `like` means satisfyin any
    # one of `#===`, `#==`, `#eql?` or `#equal?` calls.
    #
    def refute_alike(exp, act, msg=nil)
      LikeAssay.refute!(act, exp, :message=>msg, :backtrace=>caller)
    end

    #
    #
    #
    def assert_block(message="assert_block failed.", &block)
      ExecutionAssay.assert!(:message=>message, &block)
    end

    #
    # Passes if `boolean` is either `true` or `false`.
    #
    def assert_boolean(boolean, message=nil)
      BooleanAssay.assert!(boolean, :message=>message)
    end

    #
    # Passes if `boolean` is neither `true` or `false`.
    #
    def refute_boolean(boolean, message=nil)
      BooleanAssay.refute!(boolean, :message=>message)
    end

    #
    #def assert_const_defined(object, constant_name, message = nil)
    #end
    #
    #def refute_const_defined(object, constant_name, message = nil)
    #end

    # Passes if object is empty.
    #
    #   assert_empty(object)
    #
    def assert_empty(exp, msg=nil)
      EmptyAssay.assert!(exp, :message=>msg, :backtrace=>caller)
    end

    # Passes if object is not empty.
    #
    #   refute_empty(object)
    #
    def refute_empty(exp, msg=nil)
      EmptyAssay.refute!(exp, :message=>msg, :backtrace=>caller)
    end

    # Passes if expected == +actual.
    #
    # Note that the ordering of arguments is important,
    # since a helpful error message is generated when this
    # one fails that tells you the values of expected and actual.
    #
    #   assert_equal 'MY STRING', 'my string'.upcase
    #
    def assert_equal(exp, act, msg=nil)
      EqualAssay.assert!(act, exp, :message=>msg, :backtrace=>caller)
    end

    # Passes if expected != actual
    #
    #  refute_equal 'some string', 5
    #
    def refute_equal(exp, act, msg=nil)
       EqualAssay.refute!(act, exp, :message=>msg, :backtrace=>caller)
    end

    #
    #def assert_fail_assertion(message = nil) 
    #end

    #
    # Passed if object is +false+.
    #
    #   assert_false(false)
    #
    def assert_false(exp, msg=nil)
      FalseAssay.assert!(exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passed if object is not +false+.
    #
    #   refute_false(false)
    #
    def refute_false(exp, msg=nil)
      FalseAssay.refute!(exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if expected and actual are equal within absolute delta.
    #
    #   assert_in_delta 0.05, (50000.0 / 10**6), 0.00001
    #
    def assert_in_delta(exp, act, delta, msg=nil)
      WithinAssay.assert!(act, exp, delta, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if expected and actual are NOT equal within absolute delta.
    #
    #   refute_in_delta 0.05, (50000.0 / 10**6), 0.00001
    #
    def refute_in_delta(exp, act, delta, msg=nil)
      WithinAssay.refute!(act, exp, delta, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if expected and actual are equal within relative `epsilon`.
    #
    def assert_in_epsilon(exp, act, epsilon=0.001, message=nil) 
      #delta = [exp, act].min * epsilon
      CloseAssay.assert!(act, exp, epsilon, :message=>message, :backtrace=>caller)
    end

    #
    # Passes if expected and actual are NOT equal within relative `epsilon`.
    #
    def refute_in_epsilon(exp, act, epsilon=0.001, message=nil)
      #delta = [exp, act].min * epsilon
      CloseAssay.refute!(act, exp, epsilon, :message=>message, :backtrace=>caller)
    end

    #
    # Passes if `collection` contains `member`.
    #
    def assert_includes(collection, member, message=nil) 
      IncludeAssay.assert!(collection, member, :message=>message, :backtrace=>caller)
    end

    #
    # Passes if `collection` does not contain `member`.
    #
    def refute_includes(collection, member, message=nil)
      IncludeAssay.refute!(collection, member, :message=>message, :backtrace=>caller) 
    end

    #
    # Passes if object is an instance of class.
    #
    #   assert_instance_of(String, 'foo')
    #
    def assert_instance_of(cls, obj, msg=nil)
      InstanceAssay.assert!(obj, cls, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object is not an instance of class.
    #
    #   refute_instance_of(String, 500)
    #
    def refute_instance_of(cls, obj, msg=nil)
      InstanceAssay.refute!(obj, cls, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object .kind_of? klass
    #
    #   assert_kind_of(Object, 'foo')
    #
    def assert_kind_of(cls, obj, msg=nil)
      KindAssay.assert!(obj, cls, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object .kind_of? klass
    #
    #   refute_kind_of(Object, 'foo')
    #
    def refute_kind_of(cls, obj, msg=nil)
      KindAssay.refute!(obj, cls, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object matches pattern using `#=~` method.
    #
    #   assert_match(/\d+/, 'five, 6, seven')
    #
    def assert_match(pattern, string, msg=nil)
      MatchAssay.assert!(string, pattern, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object does not match pattern using `#=~` method.
    #
    #   assert_no_match(/two/, 'one 2 three')
    #
    def refute_match(pattern, string, msg=nil)
      MatchAssay.refute!(string, pattern, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object does not match pattern using `#!~` method.
    #
    #   assert_no_match(/two/, 'one 2 three')
    #
    def assert_no_match(pattern, string, msg=nil)
      NoMatchAssay.assert!(string, pattern, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object is +nil+.
    #
    #   assert_nil(nil)
    #
    def assert_nil(exp, msg=nil)
      NilAssay.assert!(exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if object is not +nil+.
    #
    #   refute_nil(true)
    #
    def refute_nil(exp, msg=nil)
      NilAssay.refute!(exp, :message=>msg, :backtrace=>caller)
    end

    # 
    # Passes if a block outputs matching test to `stdout` or `staderr`.
    #
    # This does not work _exactly_ like the original MiniTest assertion in
    # that it is an *and* condition between the two stdout and stderr, whereas
    # the original is and *or* condition.
    #
    # Note that this assertion translates into two separate underlying assertions,
    # so counts for it may be double of what one might expect.
    #
    def assert_output(stdout=nil, stderr=nil, &block)
      StdoutAssay.assert!(stdout, :backtrace=>caller, &block) if stdout
      StderrAssay.assert!(stderr, :backtrace=>caller, &block) if stderr
    end

    #
    # Passes if a block outputs matching test to `stdout` or `staderr`.
    #
    # This does not work _exactly_ like the original MiniTest assertion in
    # that it is an *and* condition between the two stdout and stderr, whereas
    # the original is and *or* condition.
    #
    # Note that this assertion translates into two separate underlying assertions,
    # so counts for it may be double of what one might expect.
    #
    def refute_output(stdout=nil, stderr=nil, &block)
      StdoutAssay.refute!(stdout, :backtrace=>caller, &block) if stdout
      StderrAssay.refute!(stderr, :backtrace=>caller, &block) if stderr
    end

    #
    # Like {#assert_output} but ensures no output.
    #
    def assert_silent(msg=nil, &block)
      SilentAssay.assert!(:message=>msg, :backtrace=>caller, &block)
    end

    #
    # Like {#refute_output} but ensures some output.
    #
    def refute_silent(msg=nil, &block)
      SilentAssay.refute!(:message=>msg, :backtrace=>caller, &block)
    end

    #
    # Passes it +predicate+ sent to object returns postively.
    #
    #   assert_predicate(10, :even?)
    #
    def assert_predicate(object, predicate, message=nil) 
      ExecutionAssay.assert!(:message=>message, :backtrace=>caller) do
        object.__send__(predicate)
      end
    end

    #
    # Passes it +predicate+ sent to object returns negatively.
    #
    #   refute_predicate(10, :odd?)
    #
    def refute_predicate(object, predicate, message=nil) 
      ExecutionAssay.refute!(:message=>message, :backtrace=>caller) do
        object.__send__(predicate)
      end
    end

    #
    # Passes if +object+ respond_to? +methods+.
    #
    #   assert_respond_to 'bugbear', :slice
    #
    def assert_respond_to(reciever, method, msg=nil)
      RespondAssay.assert!(reciever, method, :message=>msg, :backtrace=>caller)
    end
    alias_method :assert_responds_to, :assert_respond_to

    #
    # Passes if +object+ does not respond_to? +methods+.
    #
    #   refute_respond_to 'bugbear', :slice
    #
    def refute_respond_to(reciever, method, msg=nil)
      RespondAssay.refute!(reciever, method, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if +expected+ .eql? +actual+.
    #
    # Note that the ordering of arguments is important,
    # since a helpful error message is generated when this
    # one fails that tells you the values of expected and actual.
    #
    #   assert_equivalent 'MY STRING', 'my string'.upcase
    #
    def assert_equivalent(exp, act, msg=nil)
      EqualityAssay.assert!(act, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if +criterion+ is NOT equivalent to +actual+ as tested using `#eql?`.
    #
    #  refute_equivalent 'some string', 5
    #
    def refute_equivalent(criterion, act, msg=nil)
      EqualityAssay.refute!(act, criterion, :message=>msg, :backtrace=>caller)
    end

    #
    #
    #
    def assert_operator(receiver, operator, operand, message=nil) 
      ExecutionAssay.assert!(:message=>message, :backtrace=>caller) do
        receiver.__send__(operator, operand)
      end
    end

    #
    #
    #
    def refute_operator(receiver, operator, operand, message=nil) 
      ExecutionAssay.refute!(:message=>message, :backtrace=>caller) do
        receiver.__send__(operator, operand)
      end
    end

    #
    # Passes if the block raises given exception(s).
    #
    #   assert_raises RuntimeError do
    #     raise 'Boom!!!'
    #   end
    #
    def assert_raises(*exceptions, &block)
      msg = (Exception === exceptions.last ? exceptions.pop : nil)
      RaiseAssay.assert!(*exceptions, :message=>msg, :backtrace=>caller, &block)
    end

    #
    # Passes if the block *does not* raise given exception(s).
    #
    #   refute_raises RuntimeError do
    #     raise 'Boom!!!'
    #   end
    #
    def refute_raises(*exceptions, &block)
      msg = (Exception === exceptions.last ? exceptions.pop : nil)
      RaiseAssay.refute!(*exceptions, :message=>msg, :backtrace=>caller, &block)
    end

    #
    # Passes if the block yields successfully.
    #
    #   refute_nothing_raised "Couldn't do the thing" do
    #     do_the_thing
    #   end
    #
    def assert_nothing_raised(msg=nil, &block)
      RescueAssay.refute!(Exception, :message=>msg, :backtrace=>caller, &block)
    end

    #
    # Passes if the block yields successfully.
    #
    #   refute_nothing_raised "Couldn't do the thing" do
    #     do_the_thing
    #   end
    #
    def refute_nothing_raised(msg=nil, &block)
      RescueAssay.assert!(Exception, :message=>msg, :backtrace=>caller, &block)
    end

    #
    # Passes if actual is the same exact object as expected.
    #
    #   assert_same(object, object)
    #
    def assert_same(exp, act, msg=nil)
      IdentityAssay.assert!(act, exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passes if actual is not the same exact object as expected.
    #
    #   refute_same(object, other)
    #
    def refute_same(exp, act, msg=nil)
      IdentityAssay.refute!(act, exp, :message=>msg, :backtrace=>caller)
    end

    #
    #def assert_send(send_array, message=nil)
    #end

    #
    # Passes if the block throws `expected` object.
    #
    #   assert_throw :done do
    #     throw :done
    #   end
    #
    def assert_throws(expected, msg=nil, &blk)
      ThrowAssay.assert!(expected, :message=>msg, :backtrace=>caller, &blk)
    end

    #
    # Passes if the block does not throws `expected` object.
    #
    #   refute_throws :done do
    #     throw :chimp
    #   end
    #
    def refute_throws(expected, msg=nil, &blk)
      ThrowAssay.refute!(expected, :message=>msg, :backtrace=>caller, &blk)
    end

    #
    # Passed if object is +true+.
    #
    def assert_true(exp, msg=nil)
      TrueAssay.assert!(exp, :message=>msg, :backtrace=>caller)
    end

    #
    # Passed if object is not +true+.
    #
    #   refute_true(false)
    #
    def refute_true(exp, msg=nil)
      TrueAssay.refute!(exp, :message=>msg, :backtrace=>caller)
    end

  end

end
